File: apps/mobile/ios/Sources/FlagshipUI/ViewModels/HomeViewModel.swift
import Foundation
import Observation
import FlagshipAPI

@Observable
@MainActor
public final class HomeViewModel {
    public private(set) var detail: LoadingState<ServerDetailResponse> = .idle
    /// True when the LAST load failed because this device has no usable
    /// paired-session token. That includes both a missing local token and a 401
    /// from the box rejecting a stale token. The server-detail container reads
    /// this to surface the one-tap "Pair this server" affordance instead of the
    /// transient "Connecting…" placeholder (which would otherwise retry forever).
    public private(set) var needsPairing = false

    private let client: any ScreensClient
    private let podContext: String?

    public init(client: any ScreensClient, podContext: String? = nil) {
        self.client = client
        self.podContext = podContext
    }

    public func load() async {
        // Only show the skeleton on the very first attempt. On retries and
        // pull-to-refresh, keep the current state (the "Connecting…" card or
        // the last-good detail) so a transient failure doesn't flash the
        // skeleton or wipe good data underneath the user.
        if case .idle = detail { detail = .loading }
        do {
            // Mock-side: tell the in-memory client which pod fixture to
            // return. LiveScreensClient ignores this (its base URL is
            // already the right pod's daemon).
            if let mock = client as? MockScreensClient, let podContext {
                mock.podContext = podContext
            }
            let resp = try await client.serverDetail()
            detail = .loaded(resp)
            needsPairing = false
        } catch {
            // A missing OR box-rejected paired-session token is NOT transient —
            // retrying never helps until the owner pairs. A 403 is deliberately
            // excluded: it can be an operation-level authorization decision,
            // whereas the BFF's session gate reports an absent token as 401.
            switch error {
            case ScreensClientError.noSessionToken:
                needsPairing = true
            case ScreensClientError.http(status: 401, message: _):
                needsPairing = true
            default:
                needsPairing = false
            }
            // Keep showing the last successful detail on a transient refresh
            // failure; only fall back to the "Connecting…" state when we never
            // had detail to begin with.
            if case .loaded = detail { return }
            detail = .failed(HumanError.humanize(error))
        }
    }
}

File: apps/mobile/android/app/src/main/java/com/flagshipserver/app/viewmodels/HomeViewModel.kt
// Home tab + ServerDetail screen VM. Wraps the screens client's
// serverDetail() call in a LoadingState; ServerDetailScreen plus the
// Home overview both read from the same source.

package com.flagshipserver.app.viewmodels

import com.flagshipserver.app.api.ScreensClient
import com.flagshipserver.app.api.ServerDetailResponse
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

class HomeViewModel(
    private val client: ScreensClient,
    private val scope: CoroutineScope = CoroutineScope(SupervisorJob() + Dispatchers.IO),
) {
    private val _state = MutableStateFlow<LoadingState<ServerDetailResponse>>(LoadingState.Idle)
    val state: StateFlow<LoadingState<ServerDetailResponse>> = _state.asStateFlow()

    /** Single attempt — used by pull-to-refresh / the error-card retry. Only
     *  shows the skeleton on the first attempt (from Idle); keeps the last-good
     *  detail on a transient refresh failure instead of wiping it to an error. */
    fun load() = scope.launch {
        if (_state.value is LoadingState.Idle) _state.value = LoadingState.Loading
        try {
            _state.value = LoadingState.Loaded(client.serverDetail())
        } catch (t: Throwable) {
            if (_state.value !is LoadingState.Loaded) {
                _state.value = LoadingState.Failed(t.message ?: "failed to load")
            }
        }
    }

    /** Retry the BFF detail load until it lands. A box that JUST came online can
     *  take a few seconds before its daemon answers the detail BFF; a single
     *  load() would otherwise leave the page stuck (skeleton/error) until a
     *  manual refresh. Keeps showing the skeleton across retries (no error
     *  flash). Backoff 2s→15s; the launched job is cancelled when the VM's scope
     *  is torn down with the screen. */
    fun loadUntilLoaded() = scope.launch {
        if (_state.value !is LoadingState.Loaded) _state.value = LoadingState.Loading
        var delayMs = 2_000L
        while (true) {
            try {
                _state.value = LoadingState.Loaded(client.serverDetail())
                return@launch
            } catch (t: Throwable) {
                if (_state.value !is LoadingState.Loaded) _state.value = LoadingState.Loading
                delay(delayMs)
                delayMs = (delayMs * 2).coerceAtMost(15_000L)
            }
        }
    }
}
